docs/DafnyRef/UserGuide.md

```diff
@@ -156,14 +156,227 @@ exit code.
+{
+  var j := if !b then 3 else 1;
+  assert b  ==>  2 <= j;  // This assertion might not hold
+  if b {
+    assert 2 <= j;
+    i := j;
+    assert 2 <= i;
+    return;
+  }
+  i := 2;
+}
+```
+Again, now the error is only on the topmost assert, which means that we are making progress.
+Now, either the error is obvious, or we can one more time replace `j` by its value and create the assert `assert b ==> ((if !b then 3 else 1) >= 2);`
+```dafny
+method FailingPostcondition(b: bool) returns (i: int)
+  ensures 2 <= i
+{
+  assert b  ==>  2 <= (if !b then 3 else 1);  // This assertion might not hold
+  var j := if !b then 3 else 1;
+  assert b  ==>  2 <= j;
+  if b {
+    assert 2 <= j;
+    i := j;
+    assert 2 <= i;
+    return;
+  }
+  i := 2;
+}
+```
+At this point, this is pure logic. We can simplify the assumption:
+```
+b ==>  2 <= (if !b then 3 else 1)
+!b ||  (if !b then 2 <= 3 else 2 <= 1)
+!b ||  (if !b then true else false)
+!b || !b;
+!b;
+```
+Now we can understand what went wrong: When b is true, all of these formulas above are false, this is why the Dafny verifier was not able to prove them.
+In the next section, we will explain how to `move asserts up` in certain useful patterns.
+
+#### 24.8.1.3. Failing asserts cases {#sec-failing-asserts-special-cases}
+
+This list is not exhaustive but can definitely be useful to provide the next step to figure out why Dafny could not prove an assertion.
+
+ Failing assert           | Suggested rewriting
+--------------------------|---------------------------------------
+ <br>`x := Y;`<br>`assert P[x];` | `assert P[Y];`<br>`x := Y;`<br>`assert P[x];`
+ <br>`if B {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}` | `assert B ==> P;`<br>`if B {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}`
+ <br>`if B {`<br>&nbsp;&nbsp;`  ...`<br>`} else {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}` | `assert !B ==> P;`<br>`if B {`<br>&nbsp;&nbsp;`  ...`<br>`} else {`<br>&nbsp;&nbsp;`  assert P;`<br>&nbsp;&nbsp;`  ...`<br>`}`
+ <br><br>`if X {`<br>&nbsp;&nbsp;`  ...`<br>`} else {`<br>&nbsp;&nbsp;`  ...`<br>`}`<br>`assert A;` | `if X {`<br>&nbsp;&nbsp;`  ...`<br>&nbsp;&nbsp;`  assert A;`<br>`} else {`<br>&nbsp;&nbsp;`  ...`<br>&nbsp;&nbsp;`  assert A;`<br>`}`<br>`assert A;`
+ <br><br><br><br><br>`assert forall x :: Q(x);` | [`forall x`](#sec-forall-statement)<br>&nbsp;&nbsp;`  ensures Q(x)`<br>`{`<br>&nbsp;&nbsp;`  assert Q(x);`<br>`};`<br>` assert forall x :: Q(x);`
+ <br><br><br><br><br>`assert forall x :: P(x) ==> Q(x);` | [`forall x | P(x)`](#sec-forall-statement)<br>&nbsp;&nbsp;`  ensures Q(x)`<br>`{`<br>&nbsp;&nbsp;`  assert Q(x);`<br>`};`<br>` assert forall x :: P(x) ==> Q(x);`
+ <br>`assert exists x :: P(x);`<br> | `assert P(x0);`<br>`assert exists x :: P(x);`<br>for a given expression `x0`.
+ <br>`ensures exists i :: P(i);`<br> | `returns (j: int)`<br>`ensures P(j) ensures exists i :: P(i)`<br>in a lemma, so that the `j` can be computed explicitely.
```

Comment:
```suggestion
 <br>`ensures exists i :: P(i);`<br> | `returns (j: int)`<br>`ensures P(j) ensures exists i :: P(i)`<br>in a lemma, so that the `j` can be computed explicitly.
```